### PR TITLE
status entities for battery calibration & balancing

### PR DESCRIPTION
**Pull request by oliverrahner**
This PR implements #264.
It provides two new binary sensor entities that show whether battery calibration or balancing are currently active:

image

I don't like the naming much, maybe there is a better alternative?

A viable alternative to having two separate entities would be to have only one entity (like `Battery Operation Status`) with a custom enum state consisting of `Calibrating`, `Balancing`, `Normal operation`.

The current ID generation logic in this integration does not cater for the situation where **multiple entities rely on the same set of objects**, that's why I needed to "hack" my way around this by adding an actually not used object. I don't fully understand why you didn't rely on the entity `key` that is defined in `entities.py`, but there might be good reasons for that.

Last but not least: I am aware that you tried to keep a certain level of abstraction in handling objects, entities and their relation, which gets a bit contradicted by the very explicit methods `get_battery_calibration_status` and `get_battery_balancing_status`, but I couldn't come up with a better way.

### CODE REVIEW

**Comment by weltenwort** [comment]
Hi @oliverrahner, thanks for opening this PR and for providing such a useful description :heart_eyes: I'll try to set some time aside for a review within the next few days.

**Comment by weltenwort** [comment]
:information_source: I took the liberty of merging `main` into your PR branch so it uses the most recent linting action and dependency versions

**Comment by oliverrahner** [comment]
@weltenwort Is there anything I can do to have this integrated? Do you see any issues that I could solve or is it just lack of time on your end?

**Comment by weltenwort** [comment]
It's indeed just a lack of time :see_no_evil: Sorry to keep you waiting for so long. Thanks for making the changes, I'll try to prioritize looking at it this weekend :crossed_fingers:

File: custom_components/rct_power/sensor.py [hunk 1]
```diff
@@ -35,6 +37,15 @@ async def async_setup_entry(
         for entity_description in battery_sensor_entity_descriptions
     ]
 
+    battery_binary_sensor_entities = [
```
Comment (weltenwort) [inline custom_components/rct_power/sensor.py:40]: If I understand HA's component architecture correctly, setting up a binary sensor in `sensor.py` would not be correct. Instead, it should be set up as part of the `binary_sensor` platform (meaning in `binary_sensor.py`), which is being delegated to by the `__init__.py`'s loop over the `PLATFORMS` list (which would also have to be extended to include binary sensors). In light of that I actually like your alternative suggestion of introducing a single sensor that displays the state using an enum. What do you think?

**Review by weltenwort — COMMENTED** [review]
Hey @oliverrahner, sorry again for the huge delay. In general you changes make sense, but aren't quite in line with the conventions of HA sets up its platforms. You ran into this problem because you introduced a new platform (`binary_sensor`), which is meant to be setup from a `binary_sensor.py` entry point.

Your alternative suggestion for an enum-based sensor wouldn't require that.

File: custom_components/rct_power/lib/state_helpers.py [hunk 2]
```diff
@@ -60,3 +61,48 @@ def sum_api_response_values_as_state(
...
+def get_battery_calibration_status(
+    entity: BinarySensorEntity,
+    values: list[Optional[ApiResponseValue]],
+) -> Union[None, bool]:
+    if len(values) <= 0:
+        return None
+    value = values[0]
+
+    if isinstance(value, int):
+        return value & 1032 != 0
```
Comment (weltenwort) [inline custom_components/rct_power/lib/state_helpers.py:94]: Did you mean `1024` or is this actually a combination of flags? And I'd be interested to learn how you found out which flag means what.

Comment (weltenwort) [reply custom_components/rct_power/lib/state_helpers.py:94]: To answer myself: using `1032` as a binary mask is what the official app does too :shrug:

File: custom_components/rct_power/lib/entities.py [hunk 3]
```diff
@@ -191,6 +197,31 @@ def get_matching_names(expression: str):
...
+        get_native_binary_value=get_battery_calibration_status,
+    ),
+    RctPowerBinarySensorEntityDescription(
+        get_device_info=get_battery_device_info,
+        key="battery.bat_status.balancing",
+        # 'battery.status2' is not required here, this is just a hack
+        # so that this entity's generated id doesn't conflict with
+        # "battery.bat_status.calibrating"
+        # changing the id generation scheme would break existing installations
+        object_names=["battery.bat_status", "battery.status2"],
```
[the remark this answers is not in the document]
Comment (oliverrahner) [reply custom_components/rct_power/lib/entities.py:218]: Ah, I didn't think that we could just override the use of a generated id by setting it explicitly. Will change this.

File: custom_components/rct_power/sensor.py (hunk 1 above)
Comment (oliverrahner) [reply custom_components/rct_power/sensor.py:40]: TBH, I'm not fully aware of the component architecture, as I never built an integration myself. But it makes sense, even though it just seems to work in my installation without complying with the convention.
But nevertheless, I also think we should go for the enum alternative. I just thought about this after I made the current implementation.
I will refactor this and update the PR once I tested it.

**Comment by weltenwort** [comment]
Hey @oliverrahner, thanks for your willingness to refactor. Since I had the editor open I figured I could also give it a try. I opened https://github.com/oliverrahner/home-assistant-rct-power-integration/pull/1 against your main branch in case you haven't done any refactoring yourself. Please feel free to close it if you already have a better implementation.

**Comment by V4n1X** [comment]
Plans to merge these in the next release?

**Comment by weltenwort** [comment]
@oliverrahner thanks again for submitting this. I took the liberty of including your findings with some modifications in #349.

@V4n1X the new "Battery Status" entity has been released in https://github.com/weltenwort/home-assistant-rct-power-integration/releases/tag/v0.12.0.

**Comment by oliverrahner** [comment]
@weltenwort thanks for picking this up again, I lost it out of focus… much appreciated :)

**Pull request closed without merging**